gpbft/gpbft_test.go

```diff
@@ -0,0 +1,197 @@
+		em.HashJustificationOf(gpbft.COMMIT_PHASE, instance.Proposal),
+	), progressTimeout, "DECIDE step with evidence of COMMIT not reached")
+	em.RequestBroadcastFrom(1, instance.NewMessageBuilder(gpbft.Payload{
+		Step:  gpbft.DECIDE_PHASE,
+		Value: instance.Proposal,
+	}, em.GetLastSubjectBroadcast().Justification, true))
+
+	em.NextUntil(em.HasDecided(instance.ID, instance.Proposal), progressTimeout, "Did not decide on proposal")
+}
+
+func TestGPBFT_DecidesBaseOnNoStrongQuorum(t *testing.T) {
+	const progressTimeout = 2 * time.Second
+
+	em := emulator.New(t)
+	instance := &emulator.Instance{
+		ID:         0,
+		PowerTable: gpbft.NewPowerTable(),
+		Beacon:     []byte("bacon 🥓"),
+	}
+	require.NoError(t, instance.PowerTable.Add(gpbft.PowerEntry{
+		ID:     0,
+		Power:  gpbft.NewStoragePower(1),
+		PubKey: []byte("0"),
+	}, gpbft.PowerEntry{
+		ID:     1,
+		Power:  gpbft.NewStoragePower(1),
+		PubKey: []byte("1"),
+	}))
+
+	var err error
+	instance.SupplementalData.PowerTable, err = certs.MakePowerTableCID(instance.PowerTable.Entries)
+	require.NoError(t, err)
+
+	baseChain, err := gpbft.NewChain(
+		gpbft.TipSet{Epoch: 0, Key: []byte("bigbang"), PowerTable: instance.SupplementalData.PowerTable},
+	)
+	require.NoError(t, err)
+	instance.Proposal = baseChain.Extend([]byte("fish"))
+	alternativeProposal := baseChain.Extend([]byte("barreleye"))
+
+	em.SetInstance(instance)
+	em.StartInstance(instance.ID)
+
+	em.NextUntil(
+		em.HasVotedFor(gpbft.QUALITY_PHASE, instance.Proposal),
+		progressTimeout, "QUALITY step for proposal not reached")
+	em.RequestBroadcastFrom(1, instance.NewMessageBuilder(gpbft.Payload{
+		Step:  gpbft.QUALITY_PHASE,
+		Value: alternativeProposal,
+	}, nil, false))
+
+	em.NextUntil(
+		em.HasVotedFor(gpbft.PREPARE_PHASE, baseChain),
+		progressTimeout, "PREPARE step for base not reached")
+	em.RequestBroadcastFrom(1, instance.NewMessageBuilder(gpbft.Payload{
+		Step:  gpbft.PREPARE_PHASE,
+		Value: baseChain,
+	}, nil, false))
+
+	em.NextUntil(emulator.MatchesAllOf(
+		em.HasVotedFor(gpbft.COMMIT_PHASE, baseChain),
+		em.HashJustificationOf(gpbft.PREPARE_PHASE, baseChain),
+	), progressTimeout, "COMMIT step with evidence of PREPARE for base not reached")
+	em.RequestBroadcastFrom(1, instance.NewMessageBuilder(gpbft.Payload{
+		Step:  gpbft.COMMIT_PHASE,
+		Value: baseChain,
+	}, em.GetLastSubjectBroadcast().Justification, false))
+
+	em.NextUntil(emulator.MatchesAllOf(
+		em.HasVotedFor(gpbft.DECIDE_PHASE, baseChain),
+		em.HashJustificationOf(gpbft.COMMIT_PHASE, baseChain),
+	), progressTimeout, "DECIDE with evidence of COMMIT for base not reached")
+	em.RequestBroadcastFrom(1, instance.NewMessageBuilder(gpbft.Payload{
+		Step:  gpbft.DECIDE_PHASE,
+		Value: baseChain,
+	}, em.GetLastSubjectBroadcast().Justification, true))
+
+	em.NextUntil(em.HasDecided(instance.ID, baseChain), progressTimeout, "Did not decide on base")
+}
+
+func TestGPBFT_SoleParticipantDecidesOnBaseByTimeout(t *testing.T) {
```

Comment:
There aren't really any timeouts here. A participant observing all of the power vote for something should immediately progress to the next step in every case (and we want the test to show that no alarm was needed).

Reply:
Right, the name of the test is misleading. 

> we want the test to show that no alarm was needed

That's _a_ scenario yes but this test should pass even with alarm triggering. Because we cannot guarantee that (or rather the specification does not strictly require that) alarms arrive after broadcasts to self or vice versa.

Reply:
I agree these are two different scenarios, both of which should be tested. (In the alarm case, the node must still wait for the message in phases except quality and converge).